### PR TITLE
feat:pad to max seq

### PR DESCRIPTION
**Pull request by ved1beta**
# Description

Pad_to_max_sequence_len=auto support

## Motivation and Context
#1442

## How has this been tested?

unit tests 
## AI Usage Disclaimer

claude

## Summary by CodeRabbit

* **New Features**
  * Added an automatic padding mode that aligns batches to the longest dataset sample.
  * Supports automatic padding for standard and reward-model training.
  * Automatically enables compatible padding behavior when sample packing is enabled.

* **Bug Fixes**
  * Added validation to prevent automatic padding with streaming or pretraining datasets that are not map-style.

* **Tests**
  * Added coverage for automatic padding lengths and configuration validation.

### CODE REVIEW

File: src/axolotl/core/builders/causal.py [hunk 1]
```diff
@@ -385,8 +386,12 @@ def build(self, total_num_steps):
         }
         multiple = getattr(self.cfg, "pad_to_multiple_of", None) or 64
         if self.cfg.pad_to_sequence_len:
+            pad_len = self.cfg.sequence_len
+            if self.cfg.pad_to_sequence_len == "auto":
+                pad_len = self._longest_sample_len()
+                LOG.info(f"pad_to_sequence_len: auto resolved to {pad_len} tokens")
```
Comment (NanoCode012) [inline src/axolotl/core/builders/causal.py]: Can move to debug or remove

File: src/axolotl/core/builders/causal.py [hunk 2]
```diff
@@ -464,6 +469,13 @@ def build(self, total_num_steps):
 
         return trainer
 
+    def _longest_sample_len(self) -> int:
+        datasets = [d for d in (self.train_dataset, self.eval_dataset) if d]
+        if self.cfg.reward_model:
+            cols = ("chosen_ids", "rejected_ids")
+            return max(len(x) for d in datasets for c in cols for x in d[c])
+        return int(max(get_dataset_lengths(d).max() for d in datasets))
```
Comment (NanoCode012) [inline src/axolotl/core/builders/causal.py]: I can see this as being very expensive operation for larger datasets. This would break for streaming
